### ISSUE TITLE
mesos-dns fails to start

### ISSUE THREAD

**Issue by dharmit** [mesosphere/mesos-dns#236]
I am trying to start `mesos-dns` on one of the master nodes but it fails with below error.
~~~
$ sudo ./mesos-dns -config=config.json
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:17 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:20 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:20 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:23 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:23 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:23 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:23 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:24 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:24 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:28 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:28 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:28 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:28 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:29 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:29 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:29 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:29 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:29 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:29 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:29 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:29 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:32 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:32 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:36 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:36 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:36 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:36 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:36 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:36 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:40 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:40 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:40 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:40 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:41 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:41 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:41 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:41 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:41 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:41 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:41 resolver.go:331: mesos-slave-2.example.com.
ERROR: 2015/08/28 16:37:41 resolver.go:332: failed external DNS lookup of "mesos-slave-2.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:44 resolver.go:331: mesos-slave-1.example.com.
ERROR: 2015/08/28 16:37:44 resolver.go:332: failed external DNS lookup of "mesos-slave-1.example.com.": dial udp: missing port in address dns0.easydns.com.
ERROR: 2015/08/28 16:37:47 main.go:97: timed out waiting for initial ZK detection, exiting

isys at mesos-master-1 in ~ 
$ ping -qc 1 mesos-slave-2.example.com.
PING mesos-slave-2.example.com (172.19.10.214) 56(84) bytes of data.

--- mesos-slave-2.example.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.217/0.217/0.217/0.000 ms

isys at mesos-master-1 in ~ 
$ ping -qc 1 mesos-slave-1.example.com.
PING mesos-slave-1.example.com (172.19.10.213) 56(84) bytes of data.

--- mesos-slave-1.example.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.215/0.215/0.215/0.000 ms
~~~

`config.json` file:

~~~
{
  "zk": "zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos",
  "masters": ["172.19.10.111:5050","172.19.10.112:5050","172.19.10.193:5050"],
  "refreshSeconds": 60, 
  "ttl": 60, 
  "domain": "mesos",
  "ns": "ns1",
  "port": 53, 
  "resolvers": ["8.8.8.8"],
  "timeout": 5,
  "listener": "0.0.0.0",
  "SOAMname": "root.ns1.mesos",
  "SOARname": "ns1.mesos",
  "SOARefresh": 60, 
  "SOARetry":   600,
  "SOAExpire":  86400,
  "SOAMinttl": 60, 
  "dnson": true,
  "httpon": true,
  "httpport": 8123,
  "externalon": true,
  "recurseon": true
}
~~~

**Comment by dharmit** [comment]
I modified above `config.json` file to include the `resolver` for the network and tried again. Now it fails with the same ZK message:

    $ sudo ./mesos-dns -config=config.json
    ERROR: 2015/08/28 18:26:11 main.go:97: timed out waiting for initial ZK detection, exiting

I tried to telnet port 2181 on all of the nodes mentioned in `zk` field of the json file. It's working. Moreover, I am trying to deploy mesos-dns on one of the master nodes which is running ZK itself.

**Comment by n3rV3** [comment]
Make sure that your resolver DNS and the hostnames set in mesos configuration are the same. Otherwise mesos-dns complaints about it and fails. More details about failure causes can be viewed in verbose mode by adding a switch: -v=1

**Comment by dharmit** [comment]
@n3rV3 Thanks for the pointers.

> I faced an identical issue and fixed it by explicitly setting the hostname in /etc/mesos-*/hostname files.

This seemed to help me. I was testing things on one of the masters and it didn't throw any error.

So, I copied the mesos-dns binary and config.json to slaves. When I tried to start mesos-dns on slaves, it failed with similar error. Here's the verbose output:

~~~
$ sudo ./mesos-dns -config=config.json -v=1
VERBOSE: 2015/09/01 17:44:27 config.go:144: Mesos-DNS configuration:
VERBOSE: 2015/09/01 17:44:27 config.go:145:    - Masters: 172.19.10.111:5050, 172.19.10.112:5050, 172.19.10.193:5050
VERBOSE: 2015/09/01 17:44:27 config.go:146:    - Zookeeper:  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos/
VERBOSE: 2015/09/01 17:44:27 config.go:147:    - RefreshSeconds:  60
VERBOSE: 2015/09/01 17:44:27 config.go:148:    - Domain: mesos
VERBOSE: 2015/09/01 17:44:27 config.go:149:    - Listener: 0.0.0.0
VERBOSE: 2015/09/01 17:44:27 config.go:150:    - Port:  53
VERBOSE: 2015/09/01 17:44:27 config.go:151:    - DnsOn:  true
VERBOSE: 2015/09/01 17:44:27 config.go:152:    - TTL:  60
VERBOSE: 2015/09/01 17:44:27 config.go:153:    - Timeout:  5
VERBOSE: 2015/09/01 17:44:27 config.go:154:    - Resolvers: 172.16.12.201, 8.8.8.8
VERBOSE: 2015/09/01 17:44:27 config.go:155:    - ExternalOn:  true
VERBOSE: 2015/09/01 17:44:27 config.go:156:    - SOAMname: root.ns1.mesos.
VERBOSE: 2015/09/01 17:44:27 config.go:157:    - SOARname: ns1.mesos.
VERBOSE: 2015/09/01 17:44:27 config.go:158:    - SOASerial:  1441109667
VERBOSE: 2015/09/01 17:44:27 config.go:159:    - SOARefresh:  60
VERBOSE: 2015/09/01 17:44:27 config.go:160:    - SOARetry:  600
VERBOSE: 2015/09/01 17:44:27 config.go:161:    - SOAExpire:  86400
VERBOSE: 2015/09/01 17:44:27 config.go:162:    - SOAExpire:  60
VERBOSE: 2015/09/01 17:44:27 config.go:163:    - RecurseOn:  true
VERBOSE: 2015/09/01 17:44:27 config.go:164:    - HttpPort:  8123
VERBOSE: 2015/09/01 17:44:27 config.go:165:    - HttpOn:  true
VERBOSE: 2015/09/01 17:44:27 config.go:166:    - ConfigFile:  /home/isys/config.json
VERBOSE: 2015/09/01 17:44:27 config.go:167:    - EnforceRFC952:  false
VERBOSE: 2015/09/01 17:44:27 config.go:168:    - IPSources:  [mesos host]
VERBOSE: 2015/09/01 17:44:27 resolver.go:666: Starting master detector for ZK  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos/
ERROR: 2015/09/01 17:44:57 main.go:97: timed out waiting for initial ZK detection, exiting
~~~

So I checked if it's working on master. But it's failing on master with same error. I made sure to configure `/etc/mesos-*/hostname` files on masters as well as slaves.

**Comment by tsenart** [comment]
Hello @dharmit,

I'm sorry for the recent silence. I just came back from vacation today :-)
Would you please try running mesos-dns with `-v=10` so that we gather more information for debugging?

Thanks!

**Comment by dharmit** [comment]
Hi @tsenart ! 

The output is similar. Here you go:

~~~
isys at mesos-slave-1 in ~ 
$ sudo /usr/local/mesos-dns/mesos-dns -config=/usr/local/mesos-dns/config.json -v=10
VERY VERBOSE: 2015/09/01 18:52:47 config.go:144: Mesos-DNS configuration:
VERY VERBOSE: 2015/09/01 18:52:47 config.go:145:    - Masters: 172.19.10.111:5050, 172.19.10.112:5050, 172.19.10.193:5050
VERY VERBOSE: 2015/09/01 18:52:47 config.go:146:    - Zookeeper:  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos
VERY VERBOSE: 2015/09/01 18:52:47 config.go:147:    - RefreshSeconds:  60
VERY VERBOSE: 2015/09/01 18:52:47 config.go:148:    - Domain: mesos
VERY VERBOSE: 2015/09/01 18:52:47 config.go:149:    - Listener: 0.0.0.0
VERY VERBOSE: 2015/09/01 18:52:47 config.go:150:    - Port:  53
VERY VERBOSE: 2015/09/01 18:52:47 config.go:151:    - DnsOn:  true
VERY VERBOSE: 2015/09/01 18:52:47 config.go:152:    - TTL:  60
VERY VERBOSE: 2015/09/01 18:52:47 config.go:153:    - Timeout:  5
VERY VERBOSE: 2015/09/01 18:52:47 config.go:154:    - Resolvers: 172.16.12.201, 8.8.8.8
VERY VERBOSE: 2015/09/01 18:52:47 config.go:155:    - ExternalOn:  true
VERY VERBOSE: 2015/09/01 18:52:47 config.go:156:    - SOAMname: root.ns1.mesos.
VERY VERBOSE: 2015/09/01 18:52:47 config.go:157:    - SOARname: ns1.mesos.
VERY VERBOSE: 2015/09/01 18:52:47 config.go:158:    - SOASerial:  1441113767
VERY VERBOSE: 2015/09/01 18:52:47 config.go:159:    - SOARefresh:  60
VERY VERBOSE: 2015/09/01 18:52:47 config.go:160:    - SOARetry:  600
VERY VERBOSE: 2015/09/01 18:52:47 config.go:161:    - SOAExpire:  86400
VERY VERBOSE: 2015/09/01 18:52:47 config.go:162:    - SOAExpire:  60
VERY VERBOSE: 2015/09/01 18:52:47 config.go:163:    - RecurseOn:  true
VERY VERBOSE: 2015/09/01 18:52:47 config.go:164:    - HttpPort:  8123
VERY VERBOSE: 2015/09/01 18:52:47 config.go:165:    - HttpOn:  true
VERY VERBOSE: 2015/09/01 18:52:47 config.go:166:    - ConfigFile:  /usr/local/mesos-dns/config.json
VERY VERBOSE: 2015/09/01 18:52:47 config.go:167:    - EnforceRFC952:  false
VERY VERBOSE: 2015/09/01 18:52:47 config.go:168:    - IPSources:  [mesos host]
VERY VERBOSE: 2015/09/01 18:52:47 resolver.go:666: Starting master detector for ZK  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos
VERY VERBOSE: 2015/09/01 18:52:47 resolver.go:139: Warning: waiting for initial information from Zookeper.
ERROR: 2015/09/01 18:53:17 main.go:97: timed out waiting for initial ZK detection, exiting
~~~

**Comment by tsenart** [comment]
@dharmit: Indeed, that didn't give us much more information. The problem seems to be in the ZK detection code which is part of `mesos-go`. Would you please run mesos-dns with the `-logtostderr` flag? That should give us some more logs.

**Comment by dharmit** [comment]
@tsenart 

There you go:

~~~
isys at mesos-slave-1 in ~ 
$ sudo /usr/local/mesos-dns/mesos-dns -config=/usr/local/mesos-dns/config.json -v=10 -logtostderr
VERY VERBOSE: 2015/09/01 19:35:43 config.go:144: Mesos-DNS configuration:
VERY VERBOSE: 2015/09/01 19:35:43 config.go:145:    - Masters: 172.19.10.111:5050, 172.19.10.112:5050, 172.19.10.193:5050
VERY VERBOSE: 2015/09/01 19:35:43 config.go:146:    - Zookeeper:  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos
VERY VERBOSE: 2015/09/01 19:35:43 config.go:147:    - RefreshSeconds:  60
VERY VERBOSE: 2015/09/01 19:35:43 config.go:148:    - Domain: mesos
VERY VERBOSE: 2015/09/01 19:35:43 config.go:149:    - Listener: 0.0.0.0
VERY VERBOSE: 2015/09/01 19:35:43 config.go:150:    - Port:  53
VERY VERBOSE: 2015/09/01 19:35:43 config.go:151:    - DnsOn:  true
VERY VERBOSE: 2015/09/01 19:35:43 config.go:152:    - TTL:  60
VERY VERBOSE: 2015/09/01 19:35:43 config.go:153:    - Timeout:  5
VERY VERBOSE: 2015/09/01 19:35:43 config.go:154:    - Resolvers: 172.16.12.201, 8.8.8.8
VERY VERBOSE: 2015/09/01 19:35:43 config.go:155:    - ExternalOn:  true
VERY VERBOSE: 2015/09/01 19:35:43 config.go:156:    - SOAMname: root.ns1.mesos.
VERY VERBOSE: 2015/09/01 19:35:43 config.go:157:    - SOARname: ns1.mesos.
VERY VERBOSE: 2015/09/01 19:35:43 config.go:158:    - SOASerial:  1441116343
VERY VERBOSE: 2015/09/01 19:35:43 config.go:159:    - SOARefresh:  60
VERY VERBOSE: 2015/09/01 19:35:43 config.go:160:    - SOARetry:  600
VERY VERBOSE: 2015/09/01 19:35:43 config.go:161:    - SOAExpire:  86400
VERY VERBOSE: 2015/09/01 19:35:43 config.go:162:    - SOAExpire:  60
VERY VERBOSE: 2015/09/01 19:35:43 config.go:163:    - RecurseOn:  true
VERY VERBOSE: 2015/09/01 19:35:43 config.go:164:    - HttpPort:  8123
VERY VERBOSE: 2015/09/01 19:35:43 config.go:165:    - HttpOn:  true
VERY VERBOSE: 2015/09/01 19:35:43 config.go:166:    - ConfigFile:  /usr/local/mesos-dns/config.json
VERY VERBOSE: 2015/09/01 19:35:43 config.go:167:    - EnforceRFC952:  false
VERY VERBOSE: 2015/09/01 19:35:43 config.go:168:    - IPSources:  [mesos host]
VERY VERBOSE: 2015/09/01 19:35:43 resolver.go:666: Starting master detector for ZK  zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos
I0901 19:35:43.727723   19253 detect.go:79] Created new detector, watching  [172.19.10.111:2181 172.19.10.112:2181 172.19.10.193:2181] /mesos
VERY VERBOSE: 2015/09/01 19:35:43 resolver.go:139: Warning: waiting for initial information from Zookeper.
I0901 19:35:43.728311   19253 client.go:105] stateChange: from=DISCONNECTED to=REQUESTED result=true
I0901 19:35:43.728357   19253 client.go:105] stateChange: from=REQUESTED to=ATTEMPT result=true
I0901 19:35:43.728403   19253 client.go:187] Created connection object of type *zk.Conn
I0901 19:35:43.728646   19253 client.go:258] connecting to zookeeper..
I0901 19:35:43.729069   19253 client.go:261] received StateConnected
I0901 19:35:43.729117   19253 client.go:295] Watching children for path /mesos
I0901 19:35:43.729137   19253 detect.go:218] detector listener installed
I0901 19:35:43.729205   19253 detect.go:108] fetching children at path '/mesos'
W0901 19:35:43.729222   19253 detect.go:111] Unable to list children, client not connected.
W0901 19:35:43.729245   19253 client.go:320] no longer connected to server, exiting child watch
I0901 19:35:43.729262   19253 detect.go:236] resting before next detection cycle
I0901 19:35:43.729259   19253 client.go:105] stateChange: from=ATTEMPT to=CONNECTED result=true
I0901 19:35:43.729338   19253 client.go:206] zookeeper client connected
I0901 19:35:43.729351   19253 client.go:105] stateChange: from=ATTEMPT to=DISCONNECTED result=false
ERROR: 2015/09/01 19:36:13 main.go:97: timed out waiting for initial ZK detection, exiting
~~~

**Comment by tsenart** [comment]
At first sight it seems like your ZK connections are getting dropped or having packet loss. Would you be able to verify this by pinging the ZK nodes from the same machine for a while?

**Comment by tsenart** [comment]
Hmm, @jdef, could you take a look please?

**Comment by dharmit** [comment]
@tsenart ping responses are good. No packet loss.
~~~
$ ping -qc 500 172.19.10.111; ping -qc 500 172.19.10.112; ping -qc 500 172.19.10.193
PING 172.19.10.111 (172.19.10.111) 56(84) bytes of data.

--- 172.19.10.111 ping statistics ---
500 packets transmitted, 500 received, 0% packet loss, time 499000ms
rtt min/avg/max/mdev = 0.152/0.242/0.748/0.047 ms
PING 172.19.10.112 (172.19.10.112) 56(84) bytes of data.

--- 172.19.10.112 ping statistics ---
500 packets transmitted, 500 received, 0% packet loss, time 499324ms
rtt min/avg/max/mdev = 0.245/1.188/10.634/1.186 ms
PING 172.19.10.193 (172.19.10.193) 56(84) bytes of data.

--- 172.19.10.193 ping statistics ---
500 packets transmitted, 500 received, 0% packet loss, time 499435ms
rtt min/avg/max/mdev = 0.234/1.763/11.903/1.758 ms
~~~

**Comment by jdef** [comment]
@dharmit can you actually `telnet <zkhost> 2181` from your mesos-dns machine to make sure that the host:port are accessible? I'd hate to find out later that a firewall was getting in the way, and I'm somewhat distrusting of the `CONNECTED` state being reported by the ZK client.

Also, what's the max # of clients that your ZK servers are configured to accept? If you're running several frameworks, each of them may maintain multiple ongoing connections to ZK and I've definitely run into the case where ZK was configured to accept a smaller # of clients than I needed (`maxClientCnxns`, a ZK configuration setting, defaults to 10).

**Comment by jdef** [comment]
If mesos-dns is actually managing to connect to ZK and is then being disconnected for some reason, your ZK logs may indicate why.

**Comment by s-urbaniak** [comment]
I can sporadically reproduce the issue when I enter bogus entries for the mesos masters (not zookeper), i.e. "127.0.0.2:5151" or even "127.0.0.1:5151":

```
$ ./mesos-dns -logtostderr -v=10
VERY VERBOSE: 2015/09/03 20:38:00 config.go:144: Mesos-DNS configuration:
VERY VERBOSE: 2015/09/03 20:38:00 config.go:145:    - Masters: 127.0.0.1:5050, 127.0.0.2:5151
VERY VERBOSE: 2015/09/03 20:38:00 config.go:146:    - Zookeeper:  zk://127.0.0.1:2181/mesos
VERY VERBOSE: 2015/09/03 20:38:00 config.go:147:    - RefreshSeconds:  60
VERY VERBOSE: 2015/09/03 20:38:00 config.go:148:    - Domain: mesos
VERY VERBOSE: 2015/09/03 20:38:00 config.go:149:    - Listener: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:38:00 config.go:150:    - Port:  5253
VERY VERBOSE: 2015/09/03 20:38:00 config.go:151:    - DnsOn:  true
VERY VERBOSE: 2015/09/03 20:38:00 config.go:152:    - TTL:  60
VERY VERBOSE: 2015/09/03 20:38:00 config.go:153:    - Timeout:  5
VERY VERBOSE: 2015/09/03 20:38:00 config.go:154:    - Resolvers: 8.8.8.8
VERY VERBOSE: 2015/09/03 20:38:00 config.go:155:    - ExternalOn:  true
VERY VERBOSE: 2015/09/03 20:38:00 config.go:156:    - SOAMname: root.ns1.mesos.
VERY VERBOSE: 2015/09/03 20:38:00 config.go:157:    - SOARname: ns1.mesos.
VERY VERBOSE: 2015/09/03 20:38:00 config.go:158:    - SOASerial:  1441305480
VERY VERBOSE: 2015/09/03 20:38:00 config.go:159:    - SOARefresh:  60
VERY VERBOSE: 2015/09/03 20:38:00 config.go:160:    - SOARetry:  600
VERY VERBOSE: 2015/09/03 20:38:00 config.go:161:    - SOAExpire:  86400
VERY VERBOSE: 2015/09/03 20:38:00 config.go:162:    - SOAExpire:  60
VERY VERBOSE: 2015/09/03 20:38:00 config.go:163:    - RecurseOn:  true
VERY VERBOSE: 2015/09/03 20:38:00 config.go:164:    - HttpPort:  8123
VERY VERBOSE: 2015/09/03 20:38:00 config.go:165:    - HttpOn:  true
VERY VERBOSE: 2015/09/03 20:38:00 config.go:166:    - ConfigFile:  /Users/sur/src/go/src/github.com/mesosphere/mesos-dns/config.json
VERY VERBOSE: 2015/09/03 20:38:00 config.go:167:    - EnforceRFC952:  false
VERY VERBOSE: 2015/09/03 20:38:00 config.go:168:    - IPSources:  [mesos host]
VERY VERBOSE: 2015/09/03 20:38:00 resolver.go:666: Starting master detector for ZK  zk://127.0.0.1:2181/mesos
I0903 20:38:00.089406   84818 detect.go:79] Created new detector, watching  [127.0.0.1:2181] /mesos
VERY VERBOSE: 2015/09/03 20:38:00 resolver.go:139: Warning: waiting for initial information from Zookeper.
I0903 20:38:00.089485   84818 client.go:105] stateChange: from=DISCONNECTED to=REQUESTED result=true
I0903 20:38:00.089525   84818 client.go:105] stateChange: from=REQUESTED to=ATTEMPT result=true
I0903 20:38:00.089556   84818 client.go:187] Created connection object of type *zk.Conn
I0903 20:38:00.089675   84818 client.go:258] connecting to zookeeper..
I0903 20:38:00.089781   84818 client.go:261] received StateConnected
I0903 20:38:00.089799   84818 client.go:295] Watching children for path /mesos
I0903 20:38:00.089806   84818 detect.go:218] detector listener installed
I0903 20:38:00.089820   84818 detect.go:108] fetching children at path '/mesos'
W0903 20:38:00.089829   84818 detect.go:111] Unable to list children, client not connected.
W0903 20:38:00.089838   84818 client.go:320] no longer connected to server, exiting child watch
I0903 20:38:00.089848   84818 detect.go:236] resting before next detection cycle
I0903 20:38:00.089858   84818 client.go:105] stateChange: from=ATTEMPT to=CONNECTED result=true
I0903 20:38:00.089873   84818 client.go:206] zookeeper client connected
I0903 20:38:00.089878   84818 client.go:105] stateChange: from=ATTEMPT to=DISCONNECTED result=false
ERROR: 2015/09/03 20:38:30 main.go:97: timed out waiting for initial ZK detection, exiting
```

Sometimes I get a positive bootstrap situation though:

```
$ ./mesos-dns -logtostderr -v=10
VERY VERBOSE: 2015/09/03 20:39:46 config.go:144: Mesos-DNS configuration:
VERY VERBOSE: 2015/09/03 20:39:46 config.go:145:    - Masters: 127.0.0.1:5050, 127.0.0.2:5151
VERY VERBOSE: 2015/09/03 20:39:46 config.go:146:    - Zookeeper:  zk://127.0.0.1:2181/mesos
VERY VERBOSE: 2015/09/03 20:39:46 config.go:147:    - RefreshSeconds:  60
VERY VERBOSE: 2015/09/03 20:39:46 config.go:148:    - Domain: mesos
VERY VERBOSE: 2015/09/03 20:39:46 config.go:149:    - Listener: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:39:46 config.go:150:    - Port:  5253
VERY VERBOSE: 2015/09/03 20:39:46 config.go:151:    - DnsOn:  true
VERY VERBOSE: 2015/09/03 20:39:46 config.go:152:    - TTL:  60
VERY VERBOSE: 2015/09/03 20:39:46 config.go:153:    - Timeout:  5
VERY VERBOSE: 2015/09/03 20:39:46 config.go:154:    - Resolvers: 8.8.8.8
VERY VERBOSE: 2015/09/03 20:39:46 config.go:155:    - ExternalOn:  true
VERY VERBOSE: 2015/09/03 20:39:46 config.go:156:    - SOAMname: root.ns1.mesos.
VERY VERBOSE: 2015/09/03 20:39:46 config.go:157:    - SOARname: ns1.mesos.
VERY VERBOSE: 2015/09/03 20:39:46 config.go:158:    - SOASerial:  1441305586
VERY VERBOSE: 2015/09/03 20:39:46 config.go:159:    - SOARefresh:  60
VERY VERBOSE: 2015/09/03 20:39:46 config.go:160:    - SOARetry:  600
VERY VERBOSE: 2015/09/03 20:39:46 config.go:161:    - SOAExpire:  86400
VERY VERBOSE: 2015/09/03 20:39:46 config.go:162:    - SOAExpire:  60
VERY VERBOSE: 2015/09/03 20:39:46 config.go:163:    - RecurseOn:  true
VERY VERBOSE: 2015/09/03 20:39:46 config.go:164:    - HttpPort:  8123
VERY VERBOSE: 2015/09/03 20:39:46 config.go:165:    - HttpOn:  true
VERY VERBOSE: 2015/09/03 20:39:46 config.go:166:    - ConfigFile:  /Users/sur/src/go/src/github.com/mesosphere/mesos-dns/config.json
VERY VERBOSE: 2015/09/03 20:39:46 config.go:167:    - EnforceRFC952:  false
VERY VERBOSE: 2015/09/03 20:39:46 config.go:168:    - IPSources:  [mesos host]
VERY VERBOSE: 2015/09/03 20:39:46 resolver.go:666: Starting master detector for ZK  zk://127.0.0.1:2181/mesos
I0903 20:39:46.383152   85052 detect.go:79] Created new detector, watching  [127.0.0.1:2181] /mesos
VERY VERBOSE: 2015/09/03 20:39:46 resolver.go:139: Warning: waiting for initial information from Zookeper.
I0903 20:39:46.383242   85052 client.go:105] stateChange: from=DISCONNECTED to=REQUESTED result=true
I0903 20:39:46.383282   85052 client.go:105] stateChange: from=REQUESTED to=ATTEMPT result=true
I0903 20:39:46.383313   85052 client.go:187] Created connection object of type *zk.Conn
I0903 20:39:46.383445   85052 client.go:258] connecting to zookeeper..
I0903 20:39:46.383535   85052 client.go:261] received StateConnected
I0903 20:39:46.383555   85052 client.go:295] Watching children for path /mesos
I0903 20:39:46.383562   85052 detect.go:218] detector listener installed
I0903 20:39:46.383562   85052 client.go:105] stateChange: from=ATTEMPT to=CONNECTED result=true
I0903 20:39:46.383571   85052 client.go:206] zookeeper client connected
I0903 20:39:46.383578   85052 detect.go:108] fetching children at path '/mesos'
I0903 20:39:46.383579   85052 client.go:105] stateChange: from=ATTEMPT to=DISCONNECTED result=false
I0903 20:39:46.386456   85052 detect.go:268] Top node selected: 'info_0000000020'
I0903 20:39:46.386477   85052 detect.go:126] changing leader node from  -> info_0000000020
I0903 20:39:46.386631   85052 detect.go:136] detected master info: &MasterInfo{Id:*20150903-203223-16777343-5050-83503,Ip:*16777343,Port:*5050,Pid:*master@127.0.0.1:5050,Hostname:*localhost,XXX_unrecognized:[50 6 48 46 50 51 46 48],}
VERY VERBOSE: 2015/09/03 20:39:46 resolver.go:680: Updated Zookeeper info:  &MasterInfo{Id:*20150903-203223-16777343-5050-83503,Ip:*16777343,Port:*5050,Pid:*master@127.0.0.1:5050,Hostname:*localhost,XXX_unrecognized:[50 6 48 46 50 51 46 48],}
VERY VERBOSE: 2015/09/03 20:39:46 resolver.go:694: new master in Zookeeper  localhost:5050
VERY VERBOSE: 2015/09/03 20:39:46 resolver.go:142: Info: got initial information from Zookeper.
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:65: Zookeeper says the leader is:  localhost:5050
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:148: reloading from master localhost
I0903 20:39:46.386942   85052 client.go:334] rewatching children for path /mesos
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:152: Warning: master changed to localhost
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	ns1.mesos.: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	leader.mesos.: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	master.mesos.: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [SRV]	_leader._tcp.mesos.: leader.mesos.:5050
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [SRV]	_leader._udp.mesos.: leader.mesos.:5050
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	master0.mesos.: 127.0.0.1
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	master.mesos.: 127.0.0.2
VERY VERBOSE: 2015/09/03 20:39:46 generator.go:495: [A]	master1.mesos.: 127.0.0.2
VERY VERBOSE: 2015/09/03 20:39:46 logging.go:75: {MesosRequests:0 MesosSuccess:0 MesosNXDomain:0 MesosFailed:0 NonMesosRequests:0 NonMesosSuccess:0 NonMesosNXDomain:0 NonMesosFailed:0 NonMesosRecursed:0}
```

Mesos/Zk setup:

Mesos (installed via homebrew):
```
$ /usr/local/sbin/mesos-master --ip=127.0.0.1 --work_dir=. --quorum=1 --zk=zk://127.0.0.1:2181/mesos
```

Zookeeper (installed via homebrew):
```
$ zkServer start-foreground
```

@dharmit Are the mesos master nodes actually reachable on port 5050? (172.19.10.111:5050, 172.19.10.112:5050, 172.19.10.193:5050)?

**Comment by dharmit** [comment]
@jdef @s-urbaniak 

I can confirm the I can telnet from all slaves to all masters on the ports 5050 & 2181. Zookeeper instances are running on the master systems as well.

**Comment by dharmit** [comment]
What's weird is that I see timeout errors even when trying to run `mesos-dns` on the master nodes and not only on slave nodes.

**Comment by tsenart** [comment]
@jdef: Any clues?

**Comment by jdef** [comment]
The commit from Sept 3/4 probably resolves the intial bug observable in the
logs:

ERROR: 2015/08/28 16:37:17 resolver.go:332: failed external DNS lookup
of "mesos-slave-1.example.com.": dial udp: missing port in address
dns0.easydns.com.


The zk detector in mesos-go may be also be bugging out, responsible for the
initial ZK connection timeout. Or mesos itself could be misconfigured.

Can the OP please post the configuration being used for launching the mesos
masters?

**Comment by dharmit** [comment]
@jdef Below is how my `/etc/mesos-master` looks:

~~~
==> /etc/mesos-master/work_dir <==
/var/lib/mesos

==> /etc/mesos-master/hostname <==
mesos-master-1.example.com

==> /etc/mesos-master/quorum <==
2
~~~
And below is `/etc/mesos` directory:
~~~
==> /etc/mesos/zk <==
zk://172.19.10.111:2181,172.19.10.112:2181,172.19.10.193:2181/mesos
~~~
Not sure if you need any other details as a part of "configuration being used for launching the mesos master". But please let me know if you need something else.

I would like to add more details to the ZK part of  this. When I use `zkCli.sh` on masters/slaves, I am successfully able to connect and list `/mesos` using `ls` and `ls2` commands.

**Comment by jdef** [comment]
does your /etc/hosts file contain entries for your cluster nodes? since it
looks like you're using a custom domain name for the mesos hostnames, your
/etc/hosts files should match across all nodes and contain entries for each
node.

**Comment by dharmit** [comment]
Sorry for not pointing out. I have replaced `example` for the domain name of my organization. And ping works fine between the masters and slaves.

**Comment by sepiroth887** [comment]
Have you tried without specifiying the masters at all. 

Asking because i tried to add master and zk once and it didnt work for me. Though it may have been related to my setup and nginx in the way

**Comment by tsenart** [comment]
What's the status on this?

**Comment by sepiroth887** [comment]
Held up by a few other issues. Hope to get something into to a PR within the week though. Sorry

**Comment by tsenart** [comment]
@sepiroth887: Hmmm, to fix this issue in particular?

**Comment by sepiroth887** [comment]
Doh. I kinda missed what thread this was. Different issue! :(

**Comment by mindscratch** [comment]
@jdef FYI - the default for `maxClientCnxns` is 60 ([docs](http://zookeeper.apache.org/doc/r3.4.6/zookeeperAdmin.html)]

**Comment by tsenart** [comment]
@dharmit: Can we track back a little and get some more information from you please? Would you mind following our new [contribution guidelines](CONTRIBUTING.md) which asks for specific versions of things, expectations, etc. You'd just have to adapt what you have a little bit :)

**Comment by mindscratch** [comment]
What I'm seeing is mesos-dns starts up, begins working, a few minutes later I get the `timed out waiting for initial ZK detection, exiting`.

For example, at 09:01:09 I have these messages:
```
Starting master detector for ZK  zk://<ip>:2181,<ip>:2181,<ip>:2181,<ip>:2181,<ip>:2181/mesos
reloading from master <master ip>
Warning: waiting for initial information from Zookeeper.
```

At 09:05:09 I get the error `ZK watcher failed: timed out waiting for initial ZK detection, exiting`

**Comment by tsenart** [comment]
@mindscratch: What versions of Mesos-DNS, Mesos and Zookeeper are you running, on which operating system and architecture? Thanks.

**Comment by mindscratch** [comment]
@tsenart I upgraded to mesos-dns 0.3.0 but was running into issues where mesos-dns wasn't resolving hostnames (of non-mesos tasks) properly. I have `resolvers` configured, but something wasn't working quite right. I've just rolled back to mesos-dns 0.1.2, with mesos 0.24.1 and zookeeper 3.4.6 and I'm getting the error previously described.

**Comment by tsenart** [comment]
@mindscratch: Would you please try https://github.com/mesosphere/mesos-dns/releases/tag/v0.4.0 and report any bugs you find? Let us know if you're still seeing this issue in particular.

**Comment by mindscratch** [comment]
@tsenart I will as soon as I can, I have to go through a process at work to get new versions of software. Are there specific changes that you believe will resolve the issue(s) being described?

**Comment by mindscratch** [comment]
@tsenart A few other notes regarding log output from 0.3.0.

On startup I see "Starting master detector for ZK..." then `new masters detected: [master.ourapp.com:5050 192.168.3.201:5050 192.168.3.202:5050 192.168.3.203:5050]`

I have mesos-masters and zookeeper servers running on 192.168.3.{201,202,203}. `master.ourapp.com` is the hostname for 192.168.3.201.

I see every so often in the log `warning: leader "master@192.168.3.201:5050" is not in list`

I don't configure the hostname property for mesos-master, so what am I missing?

**Comment by mindscratch** [comment]
Sorry to go on and on, but also with 0.3.0 (which didn't see with 0.1.2) is:

```
error exchanging with "192.168.3.200:53": read udp 192.168.3.205:45563-><public ip>:53: i/o timeout
```

192.168.3.200 is the upstream DNS server (listed in mesos-dns config "resolvers"). What I'm thinking, based on the message is that mesos-dns talked to the "resolver", got back something about another DNS server (in this case shown as <public ip>), but the server mesos-dns is on can't get out to <public ip> (because of our networking). Does that sound reasonable? Should mesos-dns be talking to servers other than those listed in resolvers?

**Comment by tsenart** [comment]
@mindscratch: Please open a new issue to discuss these things as to not confuse everyone else on this one.

**Comment by tsenart** [comment]
@mindscratch: Or if you prefer to have a more interactive discussion, hop on IRC Freenode #mesos-dns.

**Comment by mindscratch** [comment]
@tsenart hopping on irc now, thanks.

**Comment by mindscratch** [comment]
@dharmit I've upgraded to 0.4.0 and so far I've seen no zookeeper related issues, it's working great.

**Comment by dharmit** [comment]
@tsenart I don't have this environment to reproduce the issue anymore as I changed employer.
@mindscratch Thanks for the info! I will take a note of that just in case I happen to use `mesos-dns` again.

**Comment by tsenart** [comment]
Fairly confident that improvements to the ZK code in v0.4.0 fixed this. Please re-open if this re-occurs.

**Comment by leopepe** [comment]
The same error happened with me while using mesosphere/mesos-dns:latest, after creating my own docker image based on ubuntu and installing the mesos-dns binary release 0.6.0 (as recommended on the dockerhub: https://hub.docker.com/r/mesosphere/mesos-dns/) the problem was solved.
Before that I have tested the access to the ZK from the container running mesos-dns and using the zkCli.sh I had success.
